File: clojure/src/ch1/ex116.clj
;; (load-file "filename.clj")

(defn square [x] (* x x))


(defn expt-invariant [b p a]
        (cond (= p 0) a
              (even? p) (expt-invariant (square b) (/ p 2) a)
              :else (expt-invariant b (dec p) (* a b))))


(defn expt-inv [b p]
  (expt-invariant b p 1))


(require
 'clojure.test)

(clojure.test/is (= (expt-inv 2 0) 1))
(clojure.test/is (= (expt-inv 2 1) 2))
(clojure.test/is (= (expt-inv 2 2) 4))
(clojure.test/is (= (expt-inv 2 5) 32))
(clojure.test/is (= (expt-inv 2 9) 512))
(clojure.test/is (= (expt-inv 3 4) 81))
(clojure.test/is (= (expt-inv 3 5) 243))

(println 'done)

File: scala/src/ch1/ex116.scala
def even(x: Int) = x % 2 == 0

def expt_inv(b: Int, p: Int, a: Int): Int = p match {
  case 0 => a
  case x if even(x) => expt_inv(b * b, p / 2, a)
  case _ => expt_inv(b, p - 1, a * b)
}

def expt(b: Int, p: Int) = expt_inv(b, p, 1)

assert(expt(2, 0) == 1)
assert(expt(2, 9) == 512)
assert(expt(3, 4) == 81)
assert(expt(3, 5) == 243)

println("done")
